Release/src/http/client/http_client_named_pipe.cpp

```diff
@@ -0,0 +1,466 @@
+        // As the stack frame cleans up, this will be deleted if no other references exist.
+    }
+
+private:
+
+    // Can only create on the heap using factory function.
+    named_pipe_request_context(const std::shared_ptr<_http_client_communicator> &client, const http_request &request)
+        : request_context(client, request),
+        m_bodyType(no_body),
+        m_startingPosition(std::char_traits<uint8_t>::eof()),
+        m_body_data(),
+        m_remaining_to_write(0),
+        m_readStream(request.body())
+    {
+    }
+};
+
+// named pipe client.
+class named_pipe_client : public _http_client_communicator
+{
+public:
+    static const DWORD PIPE_BUFFER_SIZE = 64 * 1024;
+
+    named_pipe_client(http::uri address, http_client_config client_config)
+        : _http_client_communicator(std::move(address), std::move(client_config))
+    {}
+
+    named_pipe_client(const named_pipe_client&) = delete;
+    named_pipe_client &operator=(const named_pipe_client&) = delete;
+
+    // Closes session.
+    ~named_pipe_client()
+    {
+        // TODO
+    }
+
+    virtual pplx::task<http_response> propagate(http_request request) override
+    {
+        auto self = std::static_pointer_cast<_http_client_communicator>(shared_from_this());
+        auto context = details::named_pipe_request_context::create_request_context(self, request);
+
+        // Use a task to externally signal the final result and completion of the task.
+        auto result_task = pplx::create_task(context->m_request_completion);
+
+        // Asynchronously send the response with the HTTP client implementation.
+        this->async_send_request(context);
+
+        return result_task;
+    }
+
+protected:
+
+    unsigned long report_failure(const utility::string_t& errorMessage)
+    {
+        // Should we log?
+        CASABLANCA_UNREFERENCED_PARAMETER(errorMessage);
+
+        return GetLastError();
+    }
+
+    // Open session and connection with the server.
+    virtual unsigned long open() override
+    {
+        const auto& config = client_config();
+
+        // TODO: Validate client config; error out on any of these conditions
+        //       * config.proxy() is enabled
+        //       * config.timeout is specified
+        //       * config.guarantee_order is specified
+        //       * config.credentials is specified
+        //       * config.validate_certificates is specified
+        //       * config.request_compressed_response is specified
+
+        // TODO: validate uri
+        auto path = uri::split_path(base_uri().path());
+
+        utility::string_t pipe_name = U("\\\\.\\pipe\\") + path[0];
+
+        m_namedPipe = CreateFile(
+            pipe_name.c_str(),
+            GENERIC_READ | GENERIC_WRITE,
+            0,
+            NULL,
+            OPEN_EXISTING,
+            0,
+            NULL);
+
+        if (m_namedPipe == INVALID_HANDLE_VALUE)
+        {
+            return report_failure(_XPLATSTR("Error opening named pipe"));
```

Comment:
Not sure but i think we will be able to show more specific error for two cases. 1 when server is not available (named pipe doesnt exist) 2. http unauthorized request 401 (access denied error)?

Reply:
That would be part of the error code... I'm checking the code to see if I can add that. Will make this explicit with a TODO.